pkg/storage/db.go

```diff
@@ -0,0 +1,283 @@
+	}
+
+	// we might be using Postgres version 10, which does not automatically propagate indices, so we do it:
+	indexName := fmt.Sprintf("%s_aws_ips_ip_ts_idx", tableName) // this is Postgres naming convention, which we don't really have to follow
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf("CREATE INDEX IF NOT EXISTS %s ON %s USING BTREE (aws_ips_ip, ts);", indexName, tableName)); err != nil {
+		return err
+	}
+
+	// this lib won't give back the last INSERTed row ID, so we don't bother with `RETURNING ...`
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`INSERT INTO %s VALUES ($1, $2, $3, $4, $5, $6)`, tableAWSEventsIPSHostnames), timestamp, isPublic, isJoin, resourceID, ipAddress, hostname); err != nil { // nolint
+		return err
+	}
+
+	return nil
+}
+
+// GetIPAddressesForTimeRange gets the audit trail of IP address and hostname changes over a specified time range
+func (db *DB) GetIPAddressesForTimeRange(ctx context.Context, start time.Time, end time.Time) ([]domain.NetworkChangeEvent, error) {
+	sqlstmt := commonQuery + fmt.Sprintf("%s.ts BETWEEN $1 AND $2;", tableAWSEventsIPSHostnames)
+	return db.runQuery(ctx, sqlstmt, start, end)
+}
+
+// GetIPAddressesForIPAddress gets the audit trail of IP Address and hostname changes for a specified IP address
+func (db *DB) GetIPAddressesForIPAddress(ctx context.Context, ipAddress string) ([]domain.NetworkChangeEvent, error) {
+	sqlstmt := commonQuery + fmt.Sprintf("%s.aws_ips_ip = $1;", tableAWSEventsIPSHostnames)
+	return db.runQuery(ctx, sqlstmt, ipAddress)
+}
+
+// GetIPAddressesForHostname gets the audit trail of IP Address and hostname changes for a specified hostname
+func (db *DB) GetIPAddressesForHostname(ctx context.Context, hostname string) ([]domain.NetworkChangeEvent, error) {
+	sqlstmt := commonQuery + fmt.Sprintf("%s.aws_hostnames_hostname = $1;", tableAWSEventsIPSHostnames)
+	return db.runQuery(ctx, sqlstmt, hostname)
+}
+
+func (db *DB) runQuery(ctx context.Context, query string, args ...interface{}) ([]domain.NetworkChangeEvent, error) {
+
+	rows, err := db.sqldb.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	networkChangeEvents := make([]domain.NetworkChangeEvent, 0)
+	for rows.Next() {
+		var row domain.NetworkChangeEvent
+		var bytes []byte
+		var hostname sql.NullString
+		if err = rows.Scan(&row.ResourceID, &row.IPAddress, &hostname, &row.IsPublic, &row.IsJoin, &row.Timestamp, &row.AccountID, &row.Region, &row.Type, &bytes); err != nil {
+			return nil, err
+		}
+		if bytes != nil {
+			var i map[string]string
+			json.Unmarshal(bytes, &i) // nolint: we already checked for nil, and the DB column is JSONB; no need for err check here
+			row.Tags = i
+		}
+		if hostname.Valid {
+			row.Hostname = hostname.String
+
+		}
+		networkChangeEvents = append(networkChangeEvents, row)
+	}
+
+	return networkChangeEvents, nil
```

Comment:
This is still missing `rows.Err()` in the return.

Reply:
nit: This could be reduced to `return networkChangeEvents, rows.Err()`. There's not obligation to ensure the left-hand value is a zero value when an error is returned. Unless otherwise documented, such as https://golang.org/pkg/io/#Reader, the general expectation is that a non-nil error means the caller must assume that any other values returned are invalid.